### PR TITLE
Properly support JUnit 5 parameters on test methods

### LINKED ISSUE #8251: ClassCastException in @ParameterizedTest: Base Class from Different Class Loader
**Describe the bug**
Running a `@QuarkusTest` fails with a `java.lang.ClassCastException`, see [MWE](https://github.com/schulzp/quarkus-classcast-mwe).

**Expected behavior**
The tests passes.

**Actual behavior**
When using `@ParameterizedTest` with `@MethodSource`, see [MWE](https://github.com/schulzp/quarkus-classcast-mwe), the test fails with the following exception:
`java.lang.ClassCastException: class org.hamcrest.core.Is cannot be cast to class org.hamcrest.Matcher (org.hamcrest.core.Is is in unnamed module of loader 'app'; org.hamcrest.Matcher is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @7188af83)`

However, if the test parameters are [mapped manually](https://github.com/schulzp/quarkus-classcast-mwe/blob/master/src/test/java/com/example/ExampleResourceTest.java#L28) (using a regular `@Test`) it passes.

**To Reproduce**
1. Check out the [MWE](https://github.com/schulzp/quarkus-classcast-mwe)
2. run `./mvnw verify`

**Configuration**
```properties
# no properties are set explicitly
```

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`:
   ```
   Darwin Kernel Version 19.3.0; root:xnu-6153.81.5~1/RELEASE_X86_64 x86_64
   ```
 - Output of `java -version`: `11.0.1, vendor: Oracle Corporation`
 - Quarkus version or git rev: `1.3.1.Final`
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): 
   ```
   Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
   Java version: 11.0.1, vendor: Oracle Corporation, runtime: [...]
   Default locale: de_DE, platform encoding: UTF-8
   OS name: "mac os x", version: "10.15.3", arch: "x86_64", family: "mac"
   ```
**Additional context**
The behavior occurred when I upgraded quarkus from 1.2.1.Final to 1.3.0.Final/1.3.1.Final.

### LINKED ISSUE #8703: QuarkusTest throws exceptions when having parameters in test function
**Describe the bug**
When having a test function with parameters injected, QuarkusTest throws `java.lang.IllegalArgumentException: argument type mismatch` exception

**Expected behavior**
Maybe it should work with parameters injected.

**Actual behavior**
```argument type mismatch
java.lang.IllegalArgumentException: argument type mismatch
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.test.junit.QuarkusTestExtension.runExtensionMethod(QuarkusTestExtension.java:433)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestMethod(QuarkusTestExtension.java:334)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:205)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:201)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:137)
```

**To Reproduce**
Steps to reproduce the behavior, simple reproduce test below:
In order to use [`VertxExtension`](https://vertx.io/docs/vertx-junit5/java/), I need to inject `VertxTestContext` in my test function:
```
@ExtendWith(VertxExtension::class)
@QuarkusTest
class simpleTest {
    @Test
    fun myTest(vertxTestContext: VertxTestContext) {
           testContext.verify { }
     }
}
```
I can create a new instance of `VertxTestContext` in the test, but will lost the benefit of its context and need to do some workaround to verify the test etc.

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`:  `Darwin Kernel Version 19.4.0`
 - Output of `java -version`: `openjdk version "11.0.6" 2020-01-14`
 - GraalVM version (if different from Java): 
 - Quarkus version or git rev: `io.quarkus:quarkus-bom:1.3.2.Final`
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`):  `6.0.1`
 - Vertx-junit5 version: `3.8.5`

### LINKED ISSUE #8978: ClassCastException when using pact-jvm Jupiter extension and @QuarkusTest
**Describe the bug**
We're using [Pact-JVM](https://github.com/DiUS/pact-jvm) to specify and test consumer driven contracts.
We implement those tests by running Junit Jupiter tests with a combination of `@QuarkusTest` and Pact-JVM Jupiter extensions.

Unfortunately since Quarkus `1.3.0.Final` those tests will fail because of `ClassCastException`. 
After a short research I've seen that those exceptions will be thrown, because the test encounters multiple ClassLoaders.
It seems that the different stages of the Pact Jupiter extension lifecycle are running with a different ClassLoader than the tests itself. So there is an incompatibility between the shared objects (parameters given from the extension into the `@Test` method).

When we use Quarkus `1.2.1.Final` everything works fine.
The issue still appears when we use Quarkus `1.4.1.Final`.

**Expected behavior**
When using Pact-JVM Jupiter Extension and `QuarkusTest` there should be no ClassCastException.

**Actual behavior**
When using Pact-JVM Jupiter Extension and `QuarkusTest` there will a ClassCastException be thrown.

**To Reproduce**

Clone this repo
`$ git clone git@github.com:tgippi/quarkus-pact-issue.git`

Run the tests either with `1.3.2.Final` or with `1.4.1.Final`:
`$ mvn -P1.3 clean test`
or
`$ mvn -P1.4 clean test`

See that there will a ClassCastException be thrown.

**Configuration**
No explicit configuration.

**Screenshots**
Shouldn't help explaining the problem.

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: 
```
Linux localhost.localdomain 5.5.5-200.fc31.x86_64 #1 SMP Wed Feb 19 23:28:07 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```
 - Output of `java -version`: 
```
openjdk version "11.0.6" 2020-01-14
OpenJDK Runtime Environment 18.9 (build 11.0.6+10)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.6+10, mixed mode, sharing)
```
 - GraalVM version (if different from Java):  No GraalVM
 - Quarkus version or git rev: Every version since `1.3.0.Final`
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): 
```
$ mvn --version
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: /opt/maven
Java version: 11.0.6, vendor: Oracle Corporation, runtime: /usr/lib/jvm/java-11-openjdk-11.0.6.10-0.fc31.x86_64
Default locale: de_DE, platform encoding: UTF-8
OS name: "linux", version: "5.5.5-200.fc31.x86_64", arch: "amd64", family: "unix"
```

### PR DESCRIPTION
**Pull request by geoand**
This is done by deep-cloning the objects into the TCCL
before actually passing them to the test

Fixes #8251, fixes #8703, fixes #8978

### CODE REVIEW

File: test-framework/junit5/pom.xml [hunk 1]
```diff
@@ -38,6 +38,13 @@
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-core</artifactId>
         </dependency>
+
+        <!-- TODO: fix -->
+        <dependency>
+            <groupId>com.thoughtworks.xstream</groupId>
+            <artifactId>xstream</artifactId>
+            <version>1.4.12</version>
```
Comment (geoand) [inline test-framework/junit5/pom.xml:46]: I will move this to the test-bom when CI completes on my fork

File: test-framework/junit5/pom.xml [hunk 2]
```diff
@@ -38,6 +38,12 @@
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-core</artifactId>
         </dependency>
+        <dependency>
+            <groupId>com.thoughtworks.xstream</groupId>
+            <artifactId>xstream</artifactId>
+            <!-- Avoid adding this to the BOM / Version has to be kept in sync with what optaplanner uses otherwise the enforcer complains -->
+            <version>1.4.11.1</version>
+        </dependency>
```
Comment (geoand) [inline test-framework/junit5/pom.xml:46]: I am assuming that having this in tests cause any problems / add any dependencies to the regular application. Is that valid?

Comment (geoand) [reply test-framework/junit5/pom.xml:46]: Seem to be

File: test-framework/junit5/src/main/java/io/quarkus/test/junit/internal/XStreamDeepClone.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import com.thoughtworks.xstream.XStream;
+
+/**
+ * Super simple cloning strategy that just serializes to XML and deserializes it using xstream
+ */
+public class XStreamDeepClone implements DeepClone {
+
+    private final XStream xStream;
+
+    public XStreamDeepClone(ClassLoader classLoader) {
```
Comment (stuartwdouglas) [inline test-framework/junit5/src/main/java/io/quarkus/test/junit/internal/XStreamDeepClone.java:12]: I don't like this, jboss marshalling has some Cloner support that is likely more appropriate to use here. 

Really though if we had support for creating an object without calling a constructor you could just use reflection to recursively clone, but adding this support requires multi release jar support and would prevent Quarkus being build with JDK8. 

Ideally we would move non-constructor based instantiation into a mini library, and then do a simple reflection based clone. I probably won't get to look at this today, but I will likely have time later this week.

Comment (geoand) [reply test-framework/junit5/src/main/java/io/quarkus/test/junit/internal/XStreamDeepClone.java:12]: Sure, that's obviously the right way to do things, but is it really warranted to spend the extra cycles for this use case?

Comment (geoand) [reply test-framework/junit5/src/main/java/io/quarkus/test/junit/internal/XStreamDeepClone.java:12]: Not to mention the various corner cases that we would need look out for as opposed to this way we seems to "just work" - although I obviously agree that it's not the cleanest solution in the world :)

Comment (geoand) [reply test-framework/junit5/src/main/java/io/quarkus/test/junit/internal/XStreamDeepClone.java:12]: Took a quick look at JBoss Marshalling and it's cloner support is just for serializable or cloneable objects

Comment (geoand) [reply test-framework/junit5/src/main/java/io/quarkus/test/junit/internal/XStreamDeepClone.java:12]: In any case, if you think it best, I'll wait until the no-constructor library is in place and then update the PR to use it - just ping me about it :)

Comment (stuartwdouglas) [reply test-framework/junit5/src/main/java/io/quarkus/test/junit/internal/XStreamDeepClone.java:12]: From memory it has an interface that you can implement to decide if an object can be cloned, and you can use to just pretend everything is serialisable. Lots of classes are also not going to work with the XML based approach.

Thinking about it wouldn't XStream also require a no-arg constructor, so a reflection based approach would not be any worse?

Comment (geoand) [reply test-framework/junit5/src/main/java/io/quarkus/test/junit/internal/XStreamDeepClone.java:12]: XStream doesn't call the constructor at all, see:

```
XStream is not calling the default constructor during deserialization.

This is, in fact, the same case as above. XStream uses the same mechanism as the JDK serialization. When using the enhanced mode with the optimized reflection API, it does not invoke the default constructor. The solution is to implement the readResolve or readObject as demonstrated with the last question.
```

from https://x-stream.github.io/faq.html.

So it should just work (and indeed in the test cases I added, the objects being passed around include ones with a no-args constructor)

Comment (geoand) [reply test-framework/junit5/src/main/java/io/quarkus/test/junit/internal/XStreamDeepClone.java:12]: FWIW, I chose this strategy because it's also used in Powemock :)

Comment (stuartwdouglas) [reply test-framework/junit5/src/main/java/io/quarkus/test/junit/internal/XStreamDeepClone.java:12]: ok, lets just go with it then

**Review by stuartwdouglas — APPROVED** [review]

File: test-framework/junit5/src/main/java/io/quarkus/test/junit/internal/XStreamDeepClone.java (hunk 3 above)
Comment (geoand) [reply test-framework/junit5/src/main/java/io/quarkus/test/junit/internal/XStreamDeepClone.java:12]: Cool :)
